File: java/com/amber/holmes/Event.java
package com.amber.holmes;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Event {
    public String eid;
    public Map<String, Object> rawEvent;
    public String tidDyn;
    public Map<String, List<String>> kgIncMap;
    public Map<String, Object[]> kgInc;

    public Event(String eid, Map<String, Object> rawEvent, String tidDyn, Map<String, List<String>> kgIncMap) {
        this.eid = eid;
        this.rawEvent = rawEvent;
        this.tidDyn = tidDyn;
        this.kgIncMap = kgIncMap;
        this.kgInc = new HashMap<>();
        genKgInc();
    }

    private void genKgInc() {
        if (kgIncMap == null) {
            assert "DENSE_BOOT".equals(tidDyn);
            return;
        }
        for (Map.Entry<String, List<String>> entry : kgIncMap.entrySet()) {
            String groupId = entry.getKey();
            List<String> fields = entry.getValue();
            Object[] fieldValues = new Object[fields.size()];
            for (int i = 0; i < fields.size(); i++) {
                fieldValues[i] = rawEvent.get(fields.get(i));
            }
            kgInc.put(groupId, fieldValues);
        }
    }
}

File: src/holmes_engine/event.py
from typing import Dict, Tuple


class Event:
    def __init__(self, eid, raw_event, tid_dyn, kg_inc_map):
        self.eid:str = eid
        self.raw_event = raw_event
        self.tid_dyn:str = tid_dyn
        self.kg_inc_map:Dict[str, Tuple] = kg_inc_map
        self.kg_inc:Dict[str, Tuple] = {}
        self.gen_kg_inc()

    def gen_kg_inc(self):
        if self.kg_inc_map is None:
            assert self.tid_dyn == "DENSE_BOOT"
            return
        for group_id in self.kg_inc_map:
            fields = self.kg_inc_map[group_id]
            fields_values = tuple([self.raw_event[name] for name in fields])
            self.kg_inc[group_id] = fields_values
